tools/tslint-rules/lightweightTokensRule.ts

```diff
@@ -0,0 +1,198 @@
+    // Symbols can be aliases of the declaration symbol. e.g. in named import specifiers.
+    // We need to resolve the aliased symbol back to the declaration symbol.
+    // tslint:disable-next-line:no-bitwise
+    if (symbol && (symbol.flags & ts.SymbolFlags.Alias) !== 0) {
+      symbol = typeChecker.getAliasedSymbol(symbol);
+    }
+
+    if (!symbol || !symbol.valueDeclaration) {
+      continue;
+    }
+
+    // Skip symbols which are explicitly disabled in the rule arguments. Some classes
+    // such as `Directionality` are always retained anyway, so there is no need to
+    // consume them through an optionally lightweight token.
+    if (ctx.options[1].includes(symbol.getName())) {
+      continue;
+    }
+
+    // No lightweight tokens are available for external symbols. Framework might
+    // add tokens in the future, but now there is no way for most symbols.
+    if (symbol.valueDeclaration.getSourceFile().isDeclarationFile) {
+      continue;
+    }
+
+    // If the containing class injects itself, then lightweight tokens will not have
+    // any effect, and we skip the lint.
+    if (symbol.valueDeclaration === node.parent) {
+      continue;
+    }
+
+    // If the parameter has a `breaking-change` JSDoc that denotes the parameter
+    // to be made `required`, skip the lightweight token failure.
+    if (hasPendingRequiredJsdoc(param)) {
+      continue;
+    }
+
+    ctx.addFailureAtNode(param.type, failureMessage);
+  }
+}
+
+/** Gets an identifier that represents the given type node's resolved value. */
+function getInjectionType(type: ts.TypeNode): ts.Identifier|null {
+  // Unwraps union types with `null`. This is supported in Angular when
+  // `@Optional` is used. The `null` union type is ignored.
+  if (ts.isUnionTypeNode(type)) {
+    const nonNullTypes = type.types.filter(t => t.kind !== ts.SyntaxKind.NullKeyword);
+    return nonNullTypes.length === 1 ? getInjectionType(nonNullTypes[0]) : null;
+  }
+  if (!ts.isTypeReferenceNode(type)) {
+    return null;
+  }
+  const unwrapEntityName = (n: ts.EntityName): ts.Identifier => {
+    if (ts.isQualifiedName(n)) {
+      return unwrapEntityName(n.right);
+    }
+    return n;
+  };
+  return unwrapEntityName(type.typeName);
+}
+
+/**
+ * Analyzes the decorators of the given parameter. Returns whether it has
+ * an `@Inject` or `@Optional` decorator applied.
+ */
+function analyzeDecorators(param: ts.ParameterDeclaration):
+    {hasInject: boolean, isOptional: boolean} {
+  if (!param.decorators) {
+    return {hasInject: false, isOptional: false};
+  }
+  let hasInject = false;
+  let isOptional = false;
+  for (const decorator of param.decorators) {
+    const expr = decorator.expression;
+    if (!ts.isCallExpression(expr) || !ts.isIdentifier(expr.expression)) {
+      continue;
+    }
+    if (expr.expression.text === 'Inject') {
+      hasInject = true;
+    } else if (expr.expression.text === 'Optional') {
```

Comment:
Shouldn't this be a separate `if`? `Inject` and `Optional` aren't mutually exclusive.

Reply:
yeah, that is a good catch